lib/Sema/CSBindings.cpp

```diff
@@ -714,12 +714,21 @@ ConstraintSystem::getPotentialBindingForRelationalConstraint(
   }
 
   // Make sure we aren't trying to equate type variables with different
-  // lvalue-binding rules.
+  // lvalue-binding rules, or convert a non-lvalue type var to an lvalue.
   if (auto otherTypeVar =
           type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
     if (typeVar->getImpl().canBindToLValue() !=
-        otherTypeVar->getImpl().canBindToLValue())
-      return None;
+        otherTypeVar->getImpl().canBindToLValue()) {
+      // An lvalue can be converted to a (possibly optional) rvalue, but we
+      // should attempt such bindings as a last resort.
+      if (constraint->getKind() >= ConstraintKind::Conversion &&
+          typeVar->getImpl().canBindToLValue()) {
+        result.FullyBound = true;
```

Comment:
We could potentially just set `FullyBound` unconditionally (i.e., just change `return None` to `result.FullyBound = true`), but I'm not sure if that would admit undesirable bindings in some cases. Going to give that a try and see if it works out...

Reply:
I feel like we should remove `lookThroughAllOptionalTypes()` from this check and do the check in a different way, maybe in `matchTypes`... I need to look again at the original example I "fixed" with this :)

Reply:
Fair enough! FWIW, delaying the bindings as long as possible rather than rejecting them outright seems to work (just pushed up the latest version), but if you can think of another good direction after some more investigation that's fine by me.